src/client.ts

```diff
@@ -5297,38 +5298,44 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
             ...res.events_before.map(mapper),
         ];
 
-        // Where the event is a thread reply (not a root) and running in MSC-enabled mode the Thread timeline only
-        // functions contiguously, so we have to jump through some hoops to get our target event in it.
-        // XXX: workaround for https://github.com/vector-im/element-meta/issues/150
-        if (Thread.hasServerSideSupport &&
-            this.supportsExperimentalThreads() &&
-            event.isRelation(THREAD_RELATION_TYPE.name)
-        ) {
-            const [, threadedEvents] = timelineSet.room.partitionThreadedEvents(events);
-            let thread = timelineSet.room.getThread(event.threadRootId);
-            if (!thread) {
-                thread = timelineSet.room.createThread(event.threadRootId, undefined, threadedEvents, true);
+        if (this.supportsExperimentalThreads()) {
+            const { threadId, shouldLiveInRoom } = timelineSet.room.eventShouldLiveIn(event);
+
+            if (!timelineSet.thread && !shouldLiveInRoom) {
+                // Thread response does not belong in this timelineSet
+                return undefined;
             }
 
-            const opts: IRelationsRequestOpts = {
-                direction: Direction.Backward,
-                limit: 50,
-            };
+            if (timelineSet.thread?.id !== threadId) {
```

Comment:
A `threadRoot` returns:
```
{
    shouldLiveInRoom: true,
    shouldLiveInThread: true,
    threadId: event.getId(),
};
```

Won't this logic prevent a `threadRoot` from showing in the main timeline? `undefined !== $abc`

Reply:
Indeed, I think this is the cause of https://github.com/vector-im/element-web/issues/22539

The if check should be `timelineSet.thread && timelineSet.thread.id !== threadId`